core/src/main/java/org/apache/iceberg/TableMetadataBuilder.java

```diff
@@ -0,0 +1,252 @@
+    this.properties = inputProperties;
+    return this;
+  }
+
+  public TableMetadataBuilder withCurrentSnapshotId(long inputCurrentSnapshotId) {
+    this.currentSnapshotId = inputCurrentSnapshotId;
+    return this;
+  }
+
+  public TableMetadataBuilder withSnapshots(List<Snapshot> inputSnapshots) {
+    this.snapshots = inputSnapshots;
+    return this;
+  }
+
+  public TableMetadataBuilder withSnapshotLog(List<HistoryEntry> inputSnapshotLog) {
+    this.snapshotLog = inputSnapshotLog;
+    return this;
+  }
+
+  public TableMetadataBuilder withPreviousFiles(List<TableMetadata.MetadataLogEntry> inputPreviousFiles) {
+    this.previousFiles = inputPreviousFiles;
+    return this;
+  }
+
+  public TableMetadataBuilder from(TableMetadata base) {
+    this.formatVersion = base.formatVersion();
+    this.uuid = base.uuid();
+    this.location = base.location();
+    this.lastSequenceNumber = base.lastSequenceNumber();
+    this.lastUpdatedMillis = base.lastUpdatedMillis();
+    this.lastColumnId = base.lastColumnId();
+    this.currentSchemaId = base.currentSchemaId();
+    this.schemas = base.schemas();
+    this.specs = base.specs();
+    this.defaultSpecId = base.defaultSpecId();
+    this.lastAssignedPartitionId = base.lastAssignedPartitionId();
+    this.defaultSortOrderId = base.defaultSortOrderId();
+    this.sortOrders = base.sortOrders();
+    this.properties = base.properties();
+    this.currentSnapshotId = base.currentSnapshotId();
+    this.snapshots = base.snapshots();
+    this.snapshotLog = base.snapshotLog();
+
+    this.baseFormatVersion = base.formatVersion();
+    this.baseFile = base.file();
+    this.baseLastUpdateMillis = base.lastUpdatedMillis();
+    this.basePreviousFiles = base.previousFiles();
+    return this;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public TableMetadata build() {
+    if (formatVersion == null) {
+      formatVersion = TableMetadata.DEFAULT_TABLE_FORMAT_VERSION;
+    } else if (formatVersion != baseFormatVersion) {
+      Preconditions.checkArgument(formatVersion <= TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION,
+          "Cannot upgrade table to unsupported format version: v%s (supported: v%s)",
+          formatVersion, TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION);
+      Preconditions.checkArgument(formatVersion >= baseFormatVersion,
+          "Cannot downgrade v%s table to v%s", baseFormatVersion, formatVersion);
+    }
+
+    if (lastSequenceNumber == null) {
+      lastSequenceNumber = TableMetadata.INITIAL_SEQUENCE_NUMBER;
+    }
+
+    if (lastUpdatedMillis == null) {
+      lastUpdatedMillis = System.currentTimeMillis();
+    }
+
+    Preconditions.checkNotNull(lastColumnId, "Last column ID not assigned for table metadata");
+    Preconditions.checkArgument(schemas != null && !schemas.isEmpty(),
+        "Table metadata must have at least 1 schema");
+    Preconditions.checkNotNull(currentSchemaId, "Table metadata must have the current schema ID");
```

Comment:
technically we can calculate the schema, spec, partition field and sort order id using a max, but for all use cases I see, we always know the actual values of these IDs and can avoid the recomputation, so I just decide to throw exception if it is not set by the caller.